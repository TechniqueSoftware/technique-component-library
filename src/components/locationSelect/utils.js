export const urlClubIdRegex = id => new RegExp(`^.*\/${id}(\/)?(#)?(((\/)?(#){1}){1}([0-9a-zA-Z])+?)*?$`, 'gm') // eslint-disable-line no-useless-escape, max-len
const urlLocationIdRegex = id => new RegExp(`^.*\/${id}\/`, 'gm') // eslint-disable-line no-useless-escape

export const getLocationIdFromUrl = (path, clubId) => {
  try {
    const [urlBeginning] = path.match(urlLocationIdRegex(clubId))
    const locationId = path.substring(urlBeginning.length).replace(/[^0-9]/g, '')
    return locationId ? parseInt(locationId, 10) : null
  } catch (err) {
    return null
  }
}
