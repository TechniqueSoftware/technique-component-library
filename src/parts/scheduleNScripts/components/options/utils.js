export const followUpTimeToObject = followUpTime => {
  const getHour = h => {
    if (h === '00') return 12
    const parsedHour = parseInt(h, 10)
    return parsedHour > 12 ? parsedHour - 12 : parsedHour
  }
  const time = followUpTime.split(':')
  const [oldHour, oldMinute] = time
  const hour = getHour(oldHour)
  const minute = parseInt(oldMinute, 10)
  const meridiem = oldHour >= 12 ? 'PM' : 'AM'
  return { hour, minute, meridiem }
}

export const objectToFollowUpTime = (hour, minute, meridiem) => {
  const parseHour = h => meridiem === 'PM' ? parseInt(h, 10) + 12 : h
  return `${parseHour(hour === 12 ? '00' : hour)}:${minute === 0 ? '00' : minute}:00`
}
