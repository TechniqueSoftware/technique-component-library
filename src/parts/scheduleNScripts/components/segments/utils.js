import { FOLLOWUP_STATUS } from './constants'

export const parseTitle = followUpStatus => {
  switch (followUpStatus) {
    case FOLLOWUP_STATUS.WEB_LEAD:
      return 'Web Lead'
    case FOLLOWUP_STATUS.APPT_BOOKED:
      return 'Appt Booked'
    case FOLLOWUP_STATUS.MEMBER:
      return 'Member'
  }
}

export const parseLocations = segmentLocations => segmentLocations.map(({ locationName, locationId }) => ({
  label: `${locationName}`,
  value: locationId
}))

export const checkPermission = userPermission => userPermission.some(
  element => element.permissionName === 'CopySegmentsAction' && element.permissionEffect === 'ALLOW')
