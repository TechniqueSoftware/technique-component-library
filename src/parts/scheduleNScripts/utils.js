import { PERMISSION_KEYS, STATUSES_WITH_SEGMENTS } from './constants'

export const checkIfSegmentStatus = (type, id) => {
  const hasSegmentation = !!STATUSES_WITH_SEGMENTS.find(s => s.type === parseInt(type, 10) && s.id === parseInt(id, 10))
  return hasSegmentation
}

const setSchedulesDayNum = schedules => schedules.reduce(
  (acc, curr) => {
    const dayNum = curr.daysTillNext + acc.dayNum
    return { schedules: acc.schedules.concat({ ...curr, dayNum }), dayNum }
  },
  { dayNum: 0, schedules: [] }
).schedules

export const getStatusSegments = (type, id, segments) => {
  if (!checkIfSegmentStatus(type, id)) return []
  const currentStatus = segments.find(
    ({ followUpType, followUpStatus }) => followUpType === parseInt(type, 10) && followUpStatus === parseInt(id, 10)
  )
  if (!currentStatus) return []
  const { followUpSegments = [], followUpSchedules = [], followUpSegmentScripts = [] } = currentStatus
  return followUpSegments.map(segment => {
    const schedulesWithScripts = followUpSchedules.reduce((acc, schedule) => {
      if (segment.segmentId !== schedule.segmentId) return acc
      return acc.concat({
        ...schedule,
        scripts: followUpSegmentScripts.filter(
          script => script.followUpSeqNum === schedule.followUpSeqNum && script.segmentId === schedule.segmentId
        )
      })
    }, [])
    return { ...segment, followUpSchedules: setSchedulesDayNum(schedulesWithScripts) }
  })
}

export const sortSegments = segments => [...segments].sort((a, b) => a.name.localeCompare(b.name))

// Formatting club-os data into something usable
export const parseSnSData = (data, followUpSchedules, followUpScripts, segmentsData, isClub) => {
  const schedulesPermissions = followUpSchedules.permissions
  const scriptsPermissions = followUpScripts.permissions
  const schedules = isClub ? followUpSchedules : followUpSchedules.schedules
  const scripts = isClub ? followUpScripts : followUpScripts.scripts

  const group = (acc, curr) => {
    const { followUpStatus, followUpType } = curr
    const key = `${followUpStatus}-${followUpType}`
    if (acc[key]) acc[key].push(curr)
    else acc[key] = [curr]
    return acc
  }

  const sch = schedules.reduce(group, {})
  const scr = scripts.reduce(group, {})

  Object.keys(sch).forEach(key => { sch[key] = setSchedulesDayNum(sch[key]) })

  Object.keys(scr).forEach(key => {
    const scripts = scr[key]
    scripts.forEach(script => {
      const schedule = sch[key]?.find(el => el.followUpSeqNum === script.followUpSeqNum)
      if (!schedule) return
      if (schedule.scripts) schedule.scripts.push(script)
      else schedule.scripts = [script]
    })
  })

  const followups = data.map(el => {
    const { type, statuses, id } = el
    const canEditSchedules = isClub ? undefined : schedulesPermissions[PERMISSION_KEYS.SCHEDULES[type]]
    const canEditScripts = isClub ? undefined : scriptsPermissions[PERMISSION_KEYS.SCRIPTS[type]]
    const newStatuses = statuses.map(st => {
      const key = `${st.id}-${type}`
      const unsortedSegments = getStatusSegments(type, st.id, segmentsData)
      const segments = sortSegments(unsortedSegments)
      return { ...st, followUpTab: id, schedules: sch[key] || [], segments }
    })
    return { ...el, statuses: newStatuses, canEditSchedules, canEditScripts }
  })

  return {
    followups,
    canEditSchedules: followups[0].canEditSchedules,
    canEditScripts: followups[0].canEditScripts
  }
}

export const getSchedulesByDay = (schedules, dayNum, order) => {
  const { earlierSch, sameDaySch, laterSch } = schedules.reduce(
    (acc, sch) => {
      if (sch.dayNum < dayNum) acc.earlierSch = acc.earlierSch.concat(sch)
      if (sch.dayNum === dayNum) acc.sameDaySch = acc.sameDaySch.concat(sch)
      if (sch.dayNum > dayNum) acc.laterSch = acc.laterSch.concat(sch)
      return acc
    },
    { earlierSch: [], sameDaySch: [], laterSch: [] }
  )
  return {
    earlierSch: order === 'first' ? earlierSch : earlierSch.concat(sameDaySch),
    laterSch: order === 'first' ? sameDaySch.concat(laterSch) : laterSch
  }
}

export const getCurrentSchedules = (segmentId, status) => {
  if (!segmentId) return status.schedules
  return status.segments.find(segment => segment.segmentId === segmentId).followUpSchedules
}

export const getStatus = (followups, type, status) => {
  const fs = followups.filter(f => f.type === type)
  if (!fs.length) return null

  let st = null
  for (let i = 0; i < fs.length; i++) {
    st = fs[i].statuses.find(s => s.id === `${status}`)
    if (st) break
  }
  return st
}

export const setFollowUpStatusValue = (followUps, status, value) => {
  const newFollowups = followUps.map(followUp => {
    if (followUp.id !== status.followUpTab) return followUp
    const newStatuses = followUp.statuses.map(st => {
      if (st.id !== status.id) return st
      return { ...st, ...value }
    })
    return { ...followUp, statuses: newStatuses }
  })
  return newFollowups
}

export const setFollowUpScripts = (followUps, { followUpType, followUpStatus }, scriptsData, segmentId) => {
  const status = getStatus(followUps, followUpType, followUpStatus)
  const currentSchedules = getCurrentSchedules(segmentId, status)
  const schedules = currentSchedules.map(schedule => {
    if (schedule.followUpSeqNum !== scriptsData.followUpSeqNum) return schedule
    return scriptsData
  })
  const statusValueToUpdate = !segmentId
    ? { schedules }
    : {
      segments: status.segments.map(
        segment => (segment.segmentId !== segmentId ? segment : { ...segment, followUpSchedules: schedules }) // eslint-disable-line max-len
      )
    }
  return setFollowUpStatusValue(followUps, status, statusValueToUpdate)
}

export const updateTouchpointInFollowups = (followUps, tp, data) => {
  const { followUpType, followUpStatus, followUpSeqNum } = tp
  const status = getStatus(followUps, followUpType, followUpStatus)
  const newSchedules = status.schedules.map(sch => {
    if (sch.followUpSeqNum !== followUpSeqNum) return sch
    const newSch = { ...sch, ...data }
    return newSch
  })
  return setFollowUpStatusValue(followUps, status, { schedules: newSchedules })
}
