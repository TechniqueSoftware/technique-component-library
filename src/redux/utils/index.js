export { default as Provider } from './provider'
export { default as TestProvider } from './testProvider'
export { default as recordSaga } from './recordSaga'
