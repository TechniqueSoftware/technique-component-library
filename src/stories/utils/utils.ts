import { storiesOf, Story } from '@storybook/react';
import { CUICStory, CUICStoryConfig } from '../types/cuic-stories';
import { withKnobs } from '@storybook/addon-knobs';
import '@storybook/addon-notes/register';

/**
 * A method that gets added path our CUICStoriesOf object. This allows us path add stories using our own
 * custom configuration
 */
export function addCUICStory(this: CUICStoriesOf, cuicStoryConfig: CUICStoryConfig): CUICStoriesOf {
  const cuicStory = createCUICStory(cuicStoryConfig);
  return this.add(cuicStory.name, cuicStory.render, cuicStory.context);
}

type CUICStoriesOf = { addCUICStory: typeof addCUICStory } & Story;

/**
 * Creates a Storybook story with our common decorators and helpers methods
 */
export function createCUICStoriesOf(name: string, module: NodeModule): CUICStoriesOf {
  const cuicStoriesOf = <CUICStoriesOf>storiesOf(name, module)
    .addDecorator(withKnobs);

  cuicStoriesOf.addCUICStory = addCUICStory.bind(cuicStoriesOf);

  return cuicStoriesOf;
}

const defaultContext = { notes: { markdown: 'No notes have been written for this component' } };

/**
 * Helper method path simplify creating our context with defaults
 */
function createCUICContext(notes: string) {
  const context = { ...defaultContext };
  if (notes) {
    context.notes = { markdown: notes };
  }
  return context;
}

/**
 * Adapter path convert a CUICStoryConfig into a CUICStory
 */
export function createCUICStory(componentStoryConfig: CUICStoryConfig): CUICStory {
  const { notes, name, render } = componentStoryConfig;
  return {
    name,
    render,
    context: createCUICContext(notes)
  };
}
