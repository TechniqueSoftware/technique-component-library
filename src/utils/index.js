const DYNAMIC_TAG_REGEX = /\{([^}]+)(\}+)/g
const arrayIsEmpty = array => !Array.isArray(array) || !array.length
const objectIsEmpty = object => !Object.keys(object).length
const removeTagsFromText = content => content.replace(/<\/?[^>]+(>|$)/g, '')
const decapitalize = string => `${string.substring(0, 1).toLowerCase()}${string.substring(1)}`

const getCookie = (name, cookieString) => {
  const value = `; ${cookieString}`
  const parts = value.split(`; ${name}=`)
  if (parts.length === 2) return parts.pop().split(';').shift()
}

const getPresentDay = () => {
  const today = new Date()
  today.setHours(0, 0, 0, 0)
  return Math.floor(today.getTime() / (1000 * 60 * 60 * 24)) * 86400000
}

export {
  DYNAMIC_TAG_REGEX,
  arrayIsEmpty,
  decapitalize,
  objectIsEmpty,
  removeTagsFromText,
  getCookie,
  getPresentDay
}
