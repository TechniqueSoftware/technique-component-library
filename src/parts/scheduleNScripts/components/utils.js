import { STATUSES_WITH_SEGMENTS } from '@parts/scheduleNScripts/constants'
// SEGMENTS
export const checkIfAttributeIsInUse = (segments, selectedAttributes, valueToCheck, segmentId) => {
  if (valueToCheck === '') return false
  const listToCheck = segments.reduce(
    (acc, segment) => {
      if (segmentId && segment.segmentId === segmentId) return acc
      return acc.concat(segment.segmentAttributes)
    },
    selectedAttributes.map(({ value }) => ({ attributeValue: value }))
  )
  return !!listToCheck.find(({ attributeValue }) => attributeValue === valueToCheck)
}

export const getSegmentStatusProp = ({ followUpType, followUpStatus }, keyword) => {
  const segmentTp = STATUSES_WITH_SEGMENTS.find(
    s => s.type === parseInt(followUpType, 10) && s.id === parseInt(followUpStatus, 10)
  )
  return segmentTp?.[keyword]
}

export const parseAttributes = (attributes, type) => attributes.reduce((acc, attr) => {
  if (!attr.value) return acc
  return acc.concat({ attributeValue: attr.value, attributeType: type, attributeId: attr.id })
}, [])

export const unparseAttributes = attributes => attributes.map(({ attributeValue, attributeId }) => ({
  name: attributeValue,
  value: attributeValue,
  id: attributeId
}))
