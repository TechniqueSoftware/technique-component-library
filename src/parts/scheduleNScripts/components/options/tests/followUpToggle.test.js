import React from 'react'
import * as redux from 'react-redux'
import { render, screen } from '@testing-library/react'
import { TestProvider as Provider } from '@redux/utils'
import userEvent from '@testing-library/user-event'

import FollowUpToggle from '../followUpToggle'

const mockhasAutoFollowUpAsTrue = () => {
  const data = {
    locationId: 4559,
    primaryAction: 2,
    secondaryAction: 3
  }

  const locationState = [{
    locationId: 4559,
    locationName: 'Philadelphia, PA'
  }]

  jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({
    global: {
      locations: locationState
    },
    scheduleNScripts: { touchpointCrud: { data, action: 1 } }
  }))
}

const mockhasAutoFollowUpAsFalse = () => {
  const data = {
    locationId: 4559
  }

  const locationState = [{
    locationId: 4559,
    locationName: 'Philadelphia, PA'
  }]

  jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb({
    global: {
      locations: locationState
    },
    scheduleNScripts: { touchpointCrud: { data, action: 1 } }
  }))
}

const props = {
  followUpToggle: 'manual',
  autoOptions: { timePicker: { hour: 12, minute: '00', meridiem: 'PM' }, minute: '' },
  setAutoOptions: jest.fn(),
  setFollowUpToggle: jest.fn(),
  isValidMinute: true
}

describe('<FollowUpToggle />', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should match snapshot', () => {
    mockhasAutoFollowUpAsTrue()
    const { asFragment } = render(
      <Provider>
        <FollowUpToggle {...props} />
      </Provider>)
    expect(asFragment()).toMatchSnapshot()
  })

  it('Renders both options manual/automatic', () => {
    mockhasAutoFollowUpAsTrue()
    render(
      <Provider>
        <FollowUpToggle {...props} />
      </Provider>)
    expect(screen.queryByText('Manual')).toBeInTheDocument()
    expect(screen.queryByText('Automatic')).toBeInTheDocument()
  })

  it('Renders automatic follow up when is selected', () => {
    mockhasAutoFollowUpAsTrue()
    render(<FollowUpToggle {...props} />)
    const optionSelected = { value: 'auto', name: 'Automatic' }
    const toggleButton = screen.queryAllByRole('radio')[1]
    userEvent.click(toggleButton)
    expect(props.setFollowUpToggle).toHaveBeenCalledWith(optionSelected.value)
  })

  it('Sets automatic if hasAutofollowUp is true', () => {
    mockhasAutoFollowUpAsTrue()
    render(
      <Provider>
        <FollowUpToggle {...props} />
      </Provider>)
    expect(props.setFollowUpToggle).toHaveBeenCalledWith('auto')
  })

  it('Not setting automatic if hasAutofollowUp is false', () => {
    mockhasAutoFollowUpAsFalse()
    render(
      <Provider>
        <FollowUpToggle {...props} />
      </Provider>)
    expect(props.setFollowUpToggle).not.toHaveBeenCalledWith('auto')
  })
}
)
