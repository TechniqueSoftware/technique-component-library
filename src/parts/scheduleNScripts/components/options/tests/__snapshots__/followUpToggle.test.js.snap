// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FollowUpToggle /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="text-center"
  >
    <div
      class="mt-3 btngroup btn-group"
      role="group"
    >
      <input
        autocomplete="off"
        checked=""
        class="btn-check"
        disabled=""
        id="radio-manual"
        name="manual-auto"
        type="radio"
        value="manual"
      />
      <label
        class="disabled btn btn-outline-primary"
        for="radio-manual"
        tabindex="0"
      >
        Manual
      </label>
      <input
        autocomplete="off"
        class="btn-check"
        disabled=""
        id="radio-auto"
        name="manual-auto"
        type="radio"
        value="auto"
      />
      <label
        class="disabled btn btn-outline-primary"
        for="radio-auto"
        tabindex="0"
      >
        Automatic
      </label>
    </div>
  </div>
</DocumentFragment>
`;
